### ISSUE TITLE
momentum sgd does not handle sparse gradients correctly

### ISSUE THREAD

**Issue by n17s** [Microsoft/CNTK#2411]
When the gradient is sparse, momentum sgd seems to skip the update but both the weights and the momentum term need updating. Here's a repro:
```python
import cntk as C
import numpy as np
import scipy.sparse

def mom_sgd(is_sparse):
    x = C.input_variable((2,), is_sparse=is_sparse)
    w = C.parameter((2,1))
    y = C.times(x,w)

    x11 = scipy.sparse.csr_matrix(np.array([1,1]).astype('f'))
    x01 = scipy.sparse.csr_matrix(np.array([0,1]).astype('f'))

    t = C.Trainer(y, y, C.momentum_sgd(y.parameters, lr=C.learning_rate_schedule(1.0, 'minibatch'), momentum=C.momentum_schedule(0.5), unit_gain=False))

    w.value = 0 * w.value
    t.train_minibatch({x:[x11]})
    t.train_minibatch({x:[x01]})
    t.train_minibatch({x:[x11]})
    return w.value

def mom(grads):
    w = 0
    m = 0
    for g in grads:
        m = 0.5 * m + g
        w = w - m
    return w
    
def bug(grads):
    w = 0
    m = 0
    for g in grads:
        if g != 0:
            m = 0.5 * m + g
            w = w - m
    return w

print('Sparse:\n', mom_sgd(is_sparse=True))
print('Dense: \n', mom_sgd(is_sparse=False))
print('py mom\n', mom([1,0,1]), mom([1,1,1]))
print('py bug\n', bug([1,0,1]), bug([1,1,1]))
```
which shows that sparse and dense disagree and the behavior of sparse is like the behavior of the python function `bug`.
